examples/fluids/qfunctions/blasius_exact.h

```diff
@@ -0,0 +1,139 @@
+//   with all their derivatives represented as a recurrence table
+// *****************************************************************************
+CEED_QFUNCTION_HELPER void ChebyshevEval(int N, const double *Tf, double x,
+    double eta_max, double *f) {
+  double dX_deta   = 2 / eta_max;
+  double table[4][3] = {
+    // Chebyshev polynomials T_0, T_1, T_2 of the first kind in (-1,1)
+    {1, x, 2*x *x - 1}, {0, 1, 4*x}, {0, 0, 4}, {0, 0, 0}
+  };
+  for (int i=0; i<4; i++) {
+    // i-th derivative of f
+    f[i] = table[i][0] * Tf[0] + table[i][1] * Tf[1] + table[i][2] * Tf[2];
+  }
+  for (int i=3; i<N; i++) {
+    // T_n(x) = 2xT_{n-1}(x) - T_{n-2}(x)
+    table[0][i%3] = 2 * x * table[0][(i-1) % 3] - table[0][(i-2)%3];
+    // Differentiate Chebyshev polynomials with the recurrence relation
+    for (int j=1; j<4; j++) {
+      // T'_{n}(x)/n = 2T_{n-1}(x) + T'_{n-2}(x)/n-2
+      table[j][i%3] = i * (2 * table[j-1][(i-1) % 3] + table[j][(i-2)%3] / (i-2));
+    }
+    for (int j=0; j<4; j++) {
+      f[j] += table[j][i%3] * Tf[i];
+    }
+  }
+  for (int i=1; i<4; i++) {
+    // Here happens the physics of the problem
+    for (int j=0; j<i; j++) f[i] *= dX_deta;
+  }
+}
+
+// *****************************************************************************
+PetscErrorCode CompressibleBlasiusResidual(SNES snes, Vec X, Vec R, void *ctx) {
+  const BlasiusContext blasius = (BlasiusContext)ctx;
+  const PetscScalar *Tf, *Th;  // Chebyshev coefficients
+  PetscScalar       *r, f[4], h[4];
+  PetscInt          N = blasius->n_cheb;
+  PetscScalar Ma = Mach(&blasius->newtonian_ctx, blasius->Tinf, blasius->Uinf),
+              Pr = Prandtl(&blasius->newtonian_ctx),
+              gamma = HeatCapacityRatio(&blasius->newtonian_ctx);
+  PetscFunctionBegin;
+  PetscCall(VecGetArrayRead(X, &Tf));
+  Th = Tf + N;
+  PetscCall(VecGetArray(R, &r));
+
+  // Compute function
+  ChebyshevEval(N, Tf, -1., blasius->eta_max, f);
+  r[0] = f[0];
+  r[1] = f[1];
+
+  // f - right end boundary condition
+  ChebyshevEval(N, Tf, 1., blasius->eta_max, f);
+  r[2] = f[1]  - 1.;
+
+  for (int i=0; i<N-3; i++) {
+    ChebyshevEval(N, Tf, blasius->X[i], blasius->eta_max, f);
+    r[3+i] = 2*f[3] + f[2] * f[0];
+    ChebyshevEval(N-1, Th, blasius->X[i], blasius->eta_max, h);
+    r[N+2+i] = h[2] + Pr * f[0] * h[1] +
+               Pr * (gamma - 1) * PetscSqr(Ma * f[2]);
+  }
+
+  // h - left end boundary condition
+  ChebyshevEval(N-1, Th, -1., blasius->eta_max, h);
+  r[N] = h[0] - blasius->T_wall / blasius->Tinf;
+
+  // h - right end boundary condition
+  ChebyshevEval(N-1, Th, 1., blasius->eta_max, h);
+  r[N+1] = h[0] - 1.;
+
+  // Restore vectors
+  PetscCall(VecRestoreArrayRead(X, &Tf));
+  PetscCall(VecRestoreArray(R, &r));
+  PetscFunctionReturn(0);
+}
+
+// *****************************************************************************
+PetscErrorCode ComputeChebyshevCoefficients(BlasiusContext blasius) {
```

Comment:
I would move this and `CompressibleBlasiusResidual` to `blasius.c`. They don't need to be called from qfunctions.